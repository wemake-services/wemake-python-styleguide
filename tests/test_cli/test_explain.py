"""Integration testing of wps explain command."""

import subprocess

import pytest


def _popen_in_shell(args: str) -> tuple[subprocess.Popen, str, str]:
    """Run command in shell."""
    # shell=True is needed for subprocess.Popen to
    # locate the installed wps command.
    process = subprocess.Popen(  # noqa: S602 (insecure shell=True)
        args,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
        shell=True,
    )
    stdin, stdout = process.communicate()
    return process, stdin, stdout


def test_command(snapshot):
    """Test that command works and formats violations as expected."""
    process, stdout, stderr = _popen_in_shell('wps explain WPS123')
    assert process.returncode == 0, (stdout, stderr)
    assert stdout == snapshot
    assert not stderr


@pytest.mark.parametrize(
    'command',
    [
        'wps explain 10000',
        'wps explain NOT_A_CODE',
        'wps explain WPS10000',
    ],
)
def test_command_on_not_found(command, snapshot):
    """Test command works when violation code is wrong."""
    process, stdout, stderr = _popen_in_shell(command)
    assert process.returncode == 1, (stdout, stderr)
    assert not stdout
    assert stderr == snapshot


def test_no_command_specified(snapshot):
    """Test command displays error message when no subcommand provided."""
    process, stdout, stderr = _popen_in_shell('wps')
    stdout, stderr = process.communicate()
    assert process.returncode != 0, (stdout, stderr)
    assert not stdout
    assert stderr == snapshot
