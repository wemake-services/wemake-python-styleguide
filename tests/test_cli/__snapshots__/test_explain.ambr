# serializer version: 1
# name: test_command
  '''
  WPS123 â€” Forbid unused variables with multiple underscores.
  
  Reasoning:
      We only use ``_`` as a special definition for an unused variable.
      Other variables are hard to read. It is unclear why would one use it.
  
  Solution:
      Rename unused variables to ``_``
      or give it some more context with an explicit name: ``_context``.
  
  Example::
  
      # Correct:
      some_element, _next_element, _ = some_tuple()
      some_element, _, _ = some_tuple()
      some_element, _ = some_tuple()
  
      # Wrong:
      some_element, _, __  = some_tuple()
  
  .. versionadded:: 0.12.0
  
  See at website: https://pyflak.es/WPS123
  
  '''
# ---
# name: test_command_on_not_found[wps explain 10000]
  '''
  Violation "10000" not found
  
  '''
# ---
# name: test_command_on_not_found[wps explain NOT_A_CODE]
  '''
  Violation "NOT_A_CODE" not found
  
  '''
# ---
# name: test_command_on_not_found[wps explain WPS10000]
  '''
  Violation "WPS10000" not found
  
  '''
# ---
# name: test_no_command_specified
  '''
  usage: wps [-h] {explain} ...
  wps: error: the following arguments are required: {explain}
  
  '''
# ---
